test/peer-churn.js

```diff
@@ -0,0 +1,404 @@
+        }
+    }
+
+    function sendAfterChurn() {
+        assert.comment('- sendAfterChurn');
+        checkRequestsTo('lucy', second, 'the second peers', client.clientChannel, assert, thenDestroySecondRound);
+    }
+
+    function thenDestroySecondRound() {
+        assert.comment('- thenDestroySecondRound');
+
+        // XXX: these don't always happen, but when they do the client retries around them
+        checkAllLogs(cluster, assert, function checkEachLog(record) {
+            if (record.msg === 'error while forwarding') {
+                var fields = record._logData.fields;
+                assert.equal(fields.error.type, 'tchannel.connection.reset',
+                    'expected connection reset while forwarding error');
+            } else {
+                assert.equal(record.msg, 'resetting connection',
+                    'expected resetting connection log');
+            }
+        });
+
+        checkTurnips();
+
+        destroyAll(second.concat(turnips).concat([client]), finish);
+    }
+
+    function finish() {
+        assert.comment('- finish');
+        assert.end();
+    }
+});
+
+function checkConnectingLog(record, assert) {
+    assert.ok([
+        'connecting peers',
+        'implementing affinity change'
+    ].indexOf(record.msg) >= 0, 'expected connection logs due to peer advertise');
+}
+
+function destroyAll(destroyables, cb) {
+    collectParallel(destroyables, destroyEach, cb);
+}
+
+function destroyEach(destroyable, i, done) {
+    destroyable.destroy(done);
+}
+
+function createRemotes(cluster, N, opts, cb) {
+    var done = CountedReadySignal(N);
+    var res = [];
+    for (var i = 0; i < N; ++i) {
+        res[i] = cluster.createRemote(opts, done.signal);
+        res[i].serverChannel.register('who', who);
+    }
+    done(cb);
+    return res;
+}
+
+function who(req, res) {
+    res.headers.as = 'raw';
+    res.sendOk('', req.channel.hostPort);
+}
+
+function checkExitsTo(cluster, serviceName, cohort, desc, assert) {
+    var cassert = CollapsedAssert();
+    for (var i = 0; i < cohort.length; ++i) {
+        var peer = cohort[i];
+        cluster.checkExitPeers(cassert, {
+            serviceName: serviceName,
+            hostPort: peer.hostPort
+        });
+    }
+    cassert.report(assert, 'expected exit peers for ' + desc);
+}
+
+/* eslint max-params: [2,8] */
+function checkRequestsTo(serviceName, cohort, desc, chan, assert, cb) {
+    var cassert = CollapsedAssert();
+    sendMany(chan, 2 * cohort.length, {
+        serviceName: serviceName,
+        timeout: 100
+    }, 'who', '', '', function sent(err, res, arg2, arg3) {
+        cassert.ifError(err, 'no unexpected error');
+        var serverHostPort = String(arg3);
+        cassert.ok(cohort.some(function isit(remote) {
+            return remote.hostPort === serverHostPort;
+        }), 'expected request to be served by one of ' + desc);
+    }, cb);
+    cassert.report(assert, 'expected requests to be served by ' + desc);
```

Comment:
:rage: you reported too early.

Please add some state to `lib/collapsed-assert` so that it throws if you call `ok()` and friends after `report()`